### PR TITLE
Add basic IPv6 framing and parsing support

### PR DESCRIPTION
**Pull request by csujedihy**
- Big-endian systems are not yet supported.
- Packets with IPv6 extensions are discarded and traced.
- ECN related bits are left zero for now.

Tested with hardcoded IPv6 address on both XDP and DPDK.

### CODE REVIEW

File: src/inc/quic_datapath.h
Comment (nibanks) [inline src/inc/quic_datapath.h:559]: I'd prefer if we could have a design that doesn't require this (though I'm not 100% against the change). I haven't looked in detail at the change yet, but I assume this is done to allocate exactly enough backfill, based on the selected address family. What if, instead, we simply always allocated enough space for the largest we'd ever need? Or do we think this would waste too much space?

File: src/platform/datapath_raw_socket.c
Comment (nibanks) [inline src/platform/datapath_raw_socket.c:241]: We follow a very explicit, 1 arg per line, format for all events in MsQuic. Please update these to match that pattern. Also, also please run `.\scripts\update-sidecar.ps1` (btw, it takes a long time).

File: src/platform/datapath_raw_socket.c
Comment (nibanks) [inline src/platform/datapath_raw_socket.c:266]: The reason I had all this "crap" before was because I noticed real IPv6 traffic had a lot of extensions before UDP in most cases. If you want to remove it for now, I'm fine with that, but I suspect it will need to come back.

File: src/platform/datapath_raw_socket.c
Comment (nibanks) [inline src/platform/datapath_raw_socket.c:304]: IMO, the byteswap is unnecessary. Just compare in wire format. I know it's a micro-optimization, but it's a trivial one to do. With all these PRs, we should keep perf at the forefront of our minds.

File: src/platform/datapath_raw_socket.c
Comment (nibanks) [reply src/platform/datapath_raw_socket.c:241]: Also, all these "unacceptable" error cases should use `LibraryErrorStatus` and pass the value in for the "status" field, probably. It might be worth having a separate event for these too; because they might be super noisy.

File: src/platform/datapath_raw_socket.c
Comment (nibanks) [inline src/platform/datapath_raw_socket.c:342]: Are these needed at the global scope? If not, push them into their specific scopes. We don't require declaring all variables at the beginning. Feel free to move them to when first used.

File: src/platform/datapath_raw_socket.c
Comment (nibanks) [inline src/platform/datapath_raw_socket.c:381]: No need to have this explicit TODO here. As you've pointed out, all of MsQuic doesn't support big-endian already. We haven't had a need to do so yet.

File: src/platform/datapath_raw_socket.c
Comment (nibanks) [inline src/platform/datapath_raw_socket.c:403]: This is interesting, but is it strictly necessary?

File: src/platform/datapath_raw_xdp.c
Comment (nibanks) [inline src/platform/datapath_raw_xdp.c:687]: Please update DPDK code as well. We are maintaining both DPDK and XDP.

File: src/inc/quic_datapath.h
Comment (thhous-msft) [reply src/inc/quic_datapath.h:559]: We would 100% need to do perf testing on this. We've attempted changes that add extra args on the send path, and it has sometimes cost 10% of performance. I feel like a few extra bytes isn't a big deal when the alternative is this invasive of a change not normally used in any other datapath.

File: src/platform/datapath_raw_socket.c
Comment (maolson-msft) [reply src/platform/datapath_raw_socket.c:241]: Is the 1-arg-per-line convention required by clog? Just curious why this convention is for only one type of macro call.

Comment (nibanks) [reply src/platform/datapath_raw_socket.c:241]: It's not a hard requirement, just our common practice that I'd prefer to keep consistent.

File: src/inc/quic_datapath.h
Comment (csujedihy) [reply src/inc/quic_datapath.h:559]: Right, we need to know how much space the header needs so that the offset for the payload can be determined. Since XDP packets are all just pre-allocated in the size of MAX_ETH_FRAME_SIZE, if we always allocate say 62 (IPv6+Eth+UDP), then we will waste 20 bytes because payload is written before headers. So, I think my current change looks very natural to me.

I am surprised that an extra argument could cause 10% perf loss. In any other datapath, this parameter should be optimized out.

File: src/platform/datapath_raw_socket.c
Comment (csujedihy) [reply src/platform/datapath_raw_socket.c:266]: Are you sure in most cases we will have extensions? I just took a packet capture for buffering a youtube video and I can't find any extensions in the QUIC flow.

Comment (nibanks) [reply src/platform/datapath_raw_socket.c:266]: Maybe it was just the local traffic in the lab. We can keep this as is for now.

File: src/platform/datapath_raw_socket.c
Comment (csujedihy) [reply src/platform/datapath_raw_socket.c:403]: You meant the flow field is necessary or the part that moves bits around? Flow ID is not strictly necessary. This field is for routers and middlebox to do QoS or load distribution. It's recommended (https://datatracker.ietf.org/doc/html/rfc6437) to have one per flow but leaving to zero works too.

Comment (nibanks) [reply src/platform/datapath_raw_socket.c:403]: I just mean, in general converting the socket pointer to a flow id. It requires an extra branch in the code that has a limited perf hit, for little or no real gain.

Comment (csujedihy) [reply src/platform/datapath_raw_socket.c:403]: Ok, I can try to skip the branch.

File: src/inc/quic_datapath.h
Comment (maolson-msft) [reply src/inc/quic_datapath.h:559]: I'm also surprised to hear that just adding args in a datapath function caused a 10% perf regression. Are you sure the regression wasn't actually from whatever logic was related to the added args?

File: src/platform/datapath_raw_xdp.c
Comment (csujedihy) [reply src/platform/datapath_raw_xdp.c:687]: DPDK updated and validated.

File: src/platform/datapath_raw.c
Comment (nibanks) [inline src/platform/datapath_raw.c:439]: There should be no need to pass family here, since the addresses are already passed in.

File: src/platform/datapath_raw_socket.c
Comment (nibanks) [inline src/platform/datapath_raw_socket.c:356]: Might there be a slight cache locality improvement to write the header from front (ethernet) to back (udp) order?

File: src/platform/datapath_raw.c
Comment (csujedihy) [reply src/platform/datapath_raw.c:439]: oh right, brainfart.

File: src/platform/datapath_raw_socket.c
Comment (csujedihy) [reply src/platform/datapath_raw_socket.c:356]: Can you explain why? Even if we try to write eth header first, the cache doesn't necessarily start with the address of the eth header. Whether all headers are in the same line is not determined by which byte we first try to access.

Comment (nibanks) [reply src/platform/datapath_raw_socket.c:356]: Just wondering.

File: src/platform/datapath_raw.c
Comment (csujedihy) [reply src/platform/datapath_raw.c]: Ok removed now

File: src/platform/datapath_raw.h
Comment (nibanks) [inline src/platform/datapath_raw.h:42]: Can this be removed now?

Comment (csujedihy) [reply src/platform/datapath_raw.h]: Sorry, I missed it. Now fixed.

**Comment by csujedihy** [comment]
Need a sign-off to unblock work from Matt.

**Review by nibanks — APPROVED** [review]

**Comment by nibanks** [comment]
Ship it, but I reserve the right to ask for more changes before it's actually merged into main.

**Comment by csujedihy** [comment]
> Ship it, but I reserve the right to ask for more changes before it's actually merged into main.

Sure thing. Thanks!

**Pull request merged**
